File: bridge/src/send-scheduler.ts
import {
  CHANNEL_WINDOW_BYTES,
  MAX_SEND_QUEUE_BYTES,
  SEAL_OVERHEAD_BYTES,
  SOCKET_INFLIGHT_BYTES,
  WINDOW_RESYNC_AGE_MS,
} from "antgrid-wire";
import type { Channel } from "./message-bus";

/** What became of a message handed to the send path. "sent" = every frame of
 *  it was sealed and written; "dropped" = at least one never will be (no
 *  session, queue cleared, stream detached, sink refused); "too-large" = the
 *  fragmenter refused it, decided synchronously at enqueue; "gated" = the
 *  stream's outbound authorization (`mayDeliver`, the remote-access switch)
 *  refused it before it reached the scheduler — the peer is fine, this
 *  machine is not talking to it. Only StreamHandle.sendTunnel produces it. */
export type SendOutcome = "sent" | "dropped" | "too-large" | "gated";

/** One frame waiting to be sealed and written: either a whole envelope or one
 *  fragment of one. Plaintext, because sealing happens at dequeue — a frame
 *  queued across a rekey must go out under the keys live at that moment. */
export interface QueuedAppFrame {
  channel: Channel;
  /** `CONTROL_STREAM_ID` for the control plane. */
  streamId: string;
  plaintext: string;
  plaintextBytes: number;
  /** Diagnostic message type, carried through to netwatch. */
  type: string;
  signal?: AbortSignal;
  authorized?: () => boolean;
  /** Fires exactly once when the frame LEAVES the queue: right after the sink
   *  wrote it ("sent") or when clear()/dropStream()/the sink dropped it. Runs
   *  inside drain(): it may only resolve a promise, never call back into the
   *  scheduler — a synchronous re-entry would hit the guard and return "idle". */
  settle?: (outcome: "sent" | "dropped") => void;
}

export interface SchedulerSink {
  /** Seal + write one frame NOW. Returns the sealed payload length that hit
   *  the wire, or null if it was dropped (no session, socket not open). */
  send(frame: QueuedAppFrame): number | null;
}

/** `idle` — nothing left to write. `held` — parked by the test seam.
 *  `blocked` — a queued head does not fit the window or the socket cap. */
export type DrainResult = "idle" | "held" | "blocked";

/**
 * Per-channel FIFO send queues with strict `control` > `preview` priority and
 * a cumulative credit window per channel plus one in-flight cap per socket.
 *
 * Pure and socket-free: the owner supplies the sink and is the only caller of
 * {@link drain}, so there is a single place where a frame reaches the wire.
 * `charge`, `uncharge` and `credit` never drain by themselves.
 *
 * Accounting is in SEALED payload bytes — the number the sender writes and the
 * receiver reads back — so a peer's cumulative credit and a relay drop report
 * are directly comparable with what was charged.
 */
export class SendScheduler {
  /** Mutable on purpose: tests shrink them. A null limit is no gate at all. */
  readonly limits = {
    window: CHANNEL_WINDOW_BYTES as number | null,
    socketCap: SOCKET_INFLIGHT_BYTES as number | null,
    maxQueuedBytes: MAX_SEND_QUEUE_BYTES,
  };
  /** Test seam: drain() returns "held" and writes nothing while true. */
  hold = false;

  private queues: Record<Channel, QueuedAppFrame[]> = { control: [], preview: [] };
  private queuedBytes: Record<Channel, number> = { control: 0, preview: 0 };
  /** Cumulative sealed bytes written on this session, queued or bypassed. */
  private sent: Record<Channel, number> = { control: 0, preview: 0 };
  /** Clamped to `sent`: over-credit is discarded rather than banked, so
   *  in-flight can never exceed one window per channel. */
  private credited: Record<Channel, number> = { control: 0, preview: 0 };
  /** The peer's raw cumulative figure, for stale/duplicate detection. */
  private lastCreditSeen: Record<Channel, number> = { control: 0, preview: 0 };
  /** `sent` as it stood when each credit arrived, oldest first. The resync
   *  rule reads the newest one old enough to be conclusive (see `credit`). */
  private anchors: Record<Channel, { at: number; sent: number }[]> = { control: [], preview: [] };
  /** Test seam for the resync clock. */
  now: () => number = Date.now;
  private draining = false;

  /** When the channel's head first failed to fit; cleared once it fits or the
   *  queue empties. Read by the owner's stall log. */
  blockedSince: Partial<Record<Channel, number>> = {};

  constructor(private sink: SchedulerSink, private log?: (msg: string) => void) {}

  unacked(ch: Channel): number {
    return Math.max(0, this.sent[ch] - this.credited[ch]);
  }

  totalUnacked(): number {
    return this.unacked("control") + this.unacked("preview");
  }

  queued(ch: Channel): { frames: number; bytes: number } {
    return { frames: this.queues[ch].length, bytes: this.queuedBytes[ch] };
  }

  /** All-or-nothing: a fragment set is never split, so a set that would push
   *  the channel past `maxQueuedBytes` is refused whole and the caller drops
   *  the message with one visible record. */
  enqueue(frames: QueuedAppFrame[]): boolean {
    if (frames.some((frame) => frame.signal?.aborted)) return false;
    if (frames.length === 0) return true;
    const adding: Record<Channel, number> = { control: 0, preview: 0 };
    for (const f of frames) adding[f.channel] += f.plaintextBytes;
    for (const ch of ["control", "preview"] as const) {
      if (this.queuedBytes[ch] + adding[ch] > this.limits.maxQueuedBytes) return false;
    }
    for (const f of frames) this.queues[f.channel].push(f);
    for (const ch of ["control", "preview"] as const) this.queuedBytes[ch] += adding[ch];
    return true;
  }

  dropAborted(): void {
    for (const channel of ["control", "preview"] as const) {
      const kept: QueuedAppFrame[] = [];
      for (const frame of this.queues[channel]) {
        if (frame.signal?.aborted) {
          this.queuedBytes[channel] -= frame.plaintextBytes;
          frame.settle?.("dropped");
        } else kept.push(frame);
      }
      this.queues[channel] = kept;
    }
  }

  drain(): DrainResult {
    // Re-entrancy guard: nothing in the current send path re-enters, but a sink
    // that ever publishes back onto a bus would otherwise interleave two loops
    // and break per-channel order.
    if (this.draining) return "idle";
    this.draining = true;
    try {
      for (;;) {
        if (this.hold) return "held";
        const next = this.pick();
        if (!next) {
          if (this.queues.control.length === 0 && this.queues.preview.length === 0) {
            this.blockedSince = {};
            return "idle";
          }
          const now = Date.now();
          for (const ch of ["control", "preview"] as const) {
            if (this.queues[ch].length > 0) this.blockedSince[ch] ??= now;
          }
          return "blocked";
        }
        delete this.blockedSince[next];
        const frame = this.queues[next].shift()!;
        this.queuedBytes[next] -= frame.plaintextBytes;
        if (frame.signal?.aborted || frame.authorized?.() === false) {
          frame.settle?.("dropped");
          continue;
        }
        const n = this.sink.send(frame);
        // A frame the sink dropped never reached the peer, so crediting it back
        // would be impossible: leave it out of the accounting entirely.
        if (n !== null) this.sent[next] += n;
        frame.settle?.(n !== null ? "sent" : "dropped");
      }
    } finally {
      this.draining = false;
    }
  }

  /** Bytes written OUTSIDE the queue (session frames): counted toward the gate,
   *  never gated by it. Charging them is what makes a relay drop report exact —
   *  the report names only a channel and a length, so bytes invisible to the
   *  accounting would un-charge something never charged. */
  charge(ch: Channel, sealedBytes: number): void {
    this.sent[ch] += sealedBytes;
  }

  /** The relay reported it discarded `bytes` of this sender's frames on `ch`.
   *  Those bytes are in `sent` and will never reach the peer's `consumed`, so
   *  without this every drop shrinks the channel's window for the session. */
  uncharge(ch: Channel, bytes: number): void {
    this.sent[ch] = Math.max(0, this.sent[ch] - bytes);
    // Nothing the peer counted can exceed what actually reached it.
    this.credited[ch] = Math.min(this.credited[ch], this.sent[ch]);
    // The anchors were taken in the old count and the discarded bytes may have
    // been written before any of them; a later resync must not present the
    // same bytes as lost a second time.
    for (const a of this.anchors[ch]) a.sent = Math.max(0, a.sent - bytes);
  }

  /** Cumulative bytes the peer says it has consumed on `ch`. Returns true iff
   *  the caller should drain: the window advanced, or a resync concluded that
   *  bytes still uncredited were lost. */
  credit(ch: Channel, consumedTotal: number): boolean {
    let moved = false;
    if (consumedTotal > this.lastCreditSeen[ch]) {
      this.lastCreditSeen[ch] = consumedTotal;
      this.credited[ch] = Math.min(consumedTotal, this.sent[ch]);
      moved = true;
    }
    const now = this.now();
    const lost = this.presumedLost(ch, now);
    if (lost > 0) {
      this.log?.(`window resync on ${ch}: ${lost} uncredited bytes presumed lost`);
      this.uncharge(ch, lost);
      moved = true;
    }
    this.anchors[ch].push({ at: now, sent: this.sent[ch] });
    return moved;
  }

  /** Bytes an anchor at least WINDOW_RESYNC_AGE_MS old saw written that the
   *  peer has still not counted. The relay delivers a channel in order and the
   *  peer credits every liveness tick, so a credit generated two ticks after a
   *  write has counted it if it ever arrived; what is still missing was
   *  discarded somewhere no drop report covered. Session frames keep both
   *  counts moving on a channel, which is why this compares against what was
   *  written rather than asking whether credits advance. A false positive
   *  costs one extra window in flight, never data. Keep in lockstep with the
   *  Dart client's send_scheduler.dart. */
  private presumedLost(ch: Channel, now: number): number {
    const anchors = this.anchors[ch];
    let newestOld = -1;
    for (let i = 0; i < anchors.length && now - anchors[i]!.at >= WINDOW_RESYNC_AGE_MS; i++) newestOld = i;
    if (newestOld < 0) return 0;
    // An older anchor can never say more than the newest conclusive one.
    anchors.splice(0, newestOld);
    return Math.max(0, anchors[0]!.sent - this.credited[ch]);
  }

  /** New session: zero every counter on both channels. Queues untouched. */
  resetWindows(): void {
    for (const ch of ["control", "preview"] as const) {
      this.sent[ch] = 0;
      this.credited[ch] = 0;
      this.lastCreditSeen[ch] = 0;
      this.anchors[ch] = [];
      delete this.blockedSince[ch];
    }
  }

  /** Drop everything; returns the dropped frames so the caller can record them. */
  clear(): QueuedAppFrame[] {
    const dropped: QueuedAppFrame[] = [];
    for (const ch of ["control", "preview"] as const) {
      dropped.push(...this.queues[ch]);
      this.queues[ch] = [];
      this.queuedBytes[ch] = 0;
      delete this.blockedSince[ch];
    }
    for (const f of dropped) f.settle?.("dropped");
    return dropped;
  }

  /** Drop queued frames for one stream (the stream detached). A detached
   *  stream's backlog must not occupy a window the live streams need. */
  dropStream(streamId: string): QueuedAppFrame[] {
    const dropped: QueuedAppFrame[] = [];
    for (const ch of ["control", "preview"] as const) {
      const kept: QueuedAppFrame[] = [];
      for (const f of this.queues[ch]) {
        if (f.streamId === streamId) {
          dropped.push(f);
          this.queuedBytes[ch] -= f.plaintextBytes;
        } else {
          kept.push(f);
        }
      }
      this.queues[ch] = kept;
      if (kept.length === 0) delete this.blockedSince[ch];
    }
    for (const f of dropped) f.settle?.("dropped");
    return dropped;
  }

  /** Control before preview, and a blocked control head never blocks preview. */
  private pick(): Channel | null {
    for (const ch of ["control", "preview"] as const) {
      const head = this.queues[ch][0];
      if (head && this.fits(head)) return ch;
    }
    return null;
  }

  private fits(f: QueuedAppFrame): boolean {
    const need = f.plaintextBytes + SEAL_OVERHEAD_BYTES;
    // The `=== 0` arms are the deadlock guards: a frame larger than a limit
    // still goes out when nothing is outstanding on it.
    const chOk =
      this.limits.window === null ||
      this.unacked(f.channel) === 0 ||
      this.unacked(f.channel) + need <= this.limits.window;
    const sockOk =
      this.limits.socketCap === null ||
      this.totalUnacked() === 0 ||
      this.totalUnacked() + need <= this.limits.socketCap;
    return chOk && sockOk;
  }
}

File: packages/antgrid_relay_client/lib/src/send_scheduler.dart
import 'dart:async';
import 'dart:math';

import 'flow.dart';

/// One frame waiting to be sealed and written: a whole envelope JSON, or one
/// fragment of one.
class QueuedAppFrame {
  QueuedAppFrame({
    required this.channel,
    required this.streamId,
    required this.plaintext,
    required this.plaintextBytes,
    this.msgType,
  });

  final String channel;
  final String streamId;
  final String plaintext;

  /// utf8 length of [plaintext] — the number the window is charged in, once the
  /// seal overhead is added. Passed in rather than computed here so a multi-MB
  /// fragment is measured once, by the encoder's own counter.
  final int plaintextBytes;

  /// Diagnostic type for the capture tap; a fragment carries its parent's.
  final String? msgType;

  /// Completes when the frame LEFT the queue — handed to the socket, or
  /// dropped. Never completes with an error: almost every caller is
  /// fire-and-forget with no handler attached, so an error here would surface
  /// as an unhandled zone error rather than as a signal. Drops stay visible
  /// where they always were, in the capture tap.
  final done = Completer<void>();
}

/// Seals and writes one frame now. Returns the sealed length that reached the
/// wire, or null if the frame was dropped before it got there.
typedef SchedulerSink = Future<int?> Function(QueuedAppFrame frame);

/// The outbound half of one E2E session: per-channel FIFO queues drained by a
/// single loop, so exactly one frame is sealed at a time and a channel's frames
/// reach the socket in the order they were handed over. Sealing at dequeue
/// rather than at enqueue means a frame queued across a rekey goes out under
/// the keys live at that moment.
///
/// [window] and [socketCap] bound the sealed bytes a sender may have in flight
/// beyond what the peer has credited, per channel and per socket. Both null
/// means no gate at all: everything queued drains as fast as the sink accepts
/// it.
///
/// Session frames (liveness, handshake, credits) never pass through here. They
/// are written directly so they overtake any backlog, and are counted against
/// the window via [charge] without being gated by it.
class SendScheduler {
  SendScheduler({
    required SchedulerSink sink,
    this.window,
    this.socketCap,
    this.maxQueuedBytes = kMaxSendQueueBytes,
    void Function(String)? log,
    DateTime Function()? clock,
  }) : _sink = sink,
       _log = log,
       clock = clock ?? DateTime.now;

  final SchedulerSink _sink;
  final void Function(String)? _log;

  /// Test seam for the resync clock.
  DateTime Function() clock;

  /// Sealed bytes allowed in flight per channel; null disables the gate.
  int? window;

  /// Sealed bytes allowed in flight across both channels; null disables it.
  int? socketCap;

  /// Per-channel cap on queued plaintext. A message whose frames would push
  /// past it is refused whole by [enqueue].
  int maxQueuedBytes;

  /// Test-only seam: the drain loop parks while true and resumes on [kick].
  /// Not part of the supported API.
  bool hold = false;

  static const _channels = ['control', 'preview'];

  final Map<String, List<QueuedAppFrame>> _queues = {
    'control': <QueuedAppFrame>[],
    'preview': <QueuedAppFrame>[],
  };
  final Map<String, int> _queuedBytes = {'control': 0, 'preview': 0};

  /// Cumulative sealed bytes written on each channel this session.
  final Map<String, int> _sent = {'control': 0, 'preview': 0};

  /// The peer's cumulative consumed count, clamped to [_sent]. Banking an
  /// over-credit would let more than one window fly at once.
  final Map<String, int> _credited = {'control': 0, 'preview': 0};

  /// The raw cumulative value last seen, so a stale or duplicated credit is
  /// recognised even after [_credited] was clamped below it.
  final Map<String, int> _lastCreditSeen = {'control': 0, 'preview': 0};

  /// [_sent] as it stood when each credit arrived, oldest first. The resync
  /// rule reads the newest one old enough to be conclusive (see [credit]).
  final Map<String, List<({DateTime at, int sent})>> _anchors = {
    'control': [],
    'preview': [],
  };

  bool _draining = false;

  /// When each channel's head first failed to fit. Cleared when it fits or the
  /// queue empties; the stall log reads it.
  final Map<String, DateTime> blockedSince = {};

  final Set<String> _stallWarned = {};

  /// Sealed bytes written on [ch] that the peer has not credited yet.
  int unacked(String ch) => max(0, _sent[ch]! - _credited[ch]!);

  int totalUnacked() => unacked('control') + unacked('preview');

  ({int frames, int bytes}) queued(String ch) =>
      (frames: _queues[ch]!.length, bytes: _queuedBytes[ch]!);

  /// Queue [frames] and start draining. All-or-nothing: a fragment set that
  /// would push its channel past [maxQueuedBytes] is refused entirely, because
  /// half a transfer on the wire is worse than none.
  bool enqueue(List<QueuedAppFrame> frames) {
    if (frames.isEmpty) return true;
    final ch = frames.first.channel;
    assert(
      frames.every((f) => f.channel == ch),
      'one call carries one message, and a message picks one channel: the '
      'queues and their byte counts are per channel',
    );
    var adding = 0;
    for (final f in frames) {
      adding += f.plaintextBytes;
    }
    if (_queuedBytes[ch]! + adding > maxQueuedBytes) return false;
    _queues[ch]!.addAll(frames);
    _queuedBytes[ch] = _queuedBytes[ch]! + adding;
    kick();
    return true;
  }

  /// Count bytes written OUTSIDE the queue — session frames, which bypass the
  /// gate but not the accounting. A peer's drop report names only a channel and
  /// a byte count, so bytes the sender never charged would un-charge bytes it
  /// did and inflate the window.
  void charge(String ch, int sealedBytes) {
    _sent[ch] = _sent[ch]! + sealedBytes;
  }

  /// The relay reported it discarded [bytes] of this sender's frames on [ch].
  /// Those bytes are in the sent total and will never reach the peer's consumed
  /// count, so without this every drop shrinks the channel's window for the
  /// rest of the session.
  void uncharge(String ch, int bytes) {
    _sent[ch] = max(0, _sent[ch]! - bytes);
    // Nothing the peer counted can exceed what actually reached it.
    _credited[ch] = min(_credited[ch]!, _sent[ch]!);
    // The anchors were taken in the old count and the discarded bytes may have
    // been written before any of them; a later resync must not present the
    // same bytes as lost a second time.
    final anchors = _anchors[ch]!;
    for (var i = 0; i < anchors.length; i++) {
      anchors[i] = (at: anchors[i].at, sent: max(0, anchors[i].sent - bytes));
    }
    kick();
  }

  /// Accept the peer's cumulative consumed count for [ch]. Returns true when
  /// the window moved — because the credit advanced, or because a resync
  /// concluded that bytes still uncredited were lost.
  ///
  /// Cumulative rather than incremental, so a credit lost in transit costs
  /// nothing: the next one carries the same ground truth.
  bool credit(String ch, int consumedTotal) {
    var moved = false;
    if (consumedTotal > _lastCreditSeen[ch]!) {
      _lastCreditSeen[ch] = consumedTotal;
      _credited[ch] = min(consumedTotal, _sent[ch]!);
      moved = true;
    }
    final now = clock();
    final lost = _presumedLost(ch, now);
    if (lost > 0) {
      _log?.call('window resync on $ch: $lost uncredited bytes presumed lost');
      uncharge(ch, lost);
      moved = true;
    }
    _anchors[ch]!.add((at: now, sent: _sent[ch]!));
    if (moved) {
      _stallWarned.remove(ch);
      kick();
    }
    return moved;
  }

  /// Bytes an anchor at least [kWindowResyncAgeMs] old saw written that the
  /// peer has still not counted. The relay delivers a channel in order and the
  /// peer credits every liveness tick, so a credit generated two ticks after a
  /// write has counted it if it ever arrived; what is still missing was
  /// discarded somewhere no drop report covered. Session frames keep both
  /// counts moving on a channel, which is why this compares against what was
  /// written rather than asking whether credits advance. A false positive
  /// costs one extra window in flight, never data. Keep in lockstep with the
  /// bridge's send-scheduler.ts.
  int _presumedLost(String ch, DateTime now) {
    final anchors = _anchors[ch]!;
    var newestOld = -1;
    for (
      var i = 0;
      i < anchors.length &&
          now.difference(anchors[i].at).inMilliseconds >= kWindowResyncAgeMs;
      i++
    ) {
      newestOld = i;
    }
    if (newestOld < 0) return 0;
    // An older anchor can never say more than the newest conclusive one.
    anchors.removeRange(0, newestOld);
    return max(0, anchors[0].sent - _credited[ch]!);
  }

  /// Forget every counter on both channels — a new session credits from zero.
  /// The queues are untouched: frames held across a rekey are still owed.
  void resetWindows() {
    for (final ch in _channels) {
      _sent[ch] = 0;
      _credited[ch] = 0;
      _lastCreditSeen[ch] = 0;
      _anchors[ch]!.clear();
    }
    blockedSince.clear();
    _stallWarned.clear();
  }

  /// Drop everything queued and return it, so the caller can record the drops.
  /// Every returned frame's `done` is completed, never failed.
  List<QueuedAppFrame> clear() {
    final dropped = <QueuedAppFrame>[];
    for (final ch in _channels) {
      dropped.addAll(_queues[ch]!);
      _queues[ch]!.clear();
      _queuedBytes[ch] = 0;
    }
    blockedSince.clear();
    _complete(dropped);
    return dropped;
  }

  /// Drop the queued frames of one stream — it detached, so its backlog must
  /// not sit in a window the rest of the session needs. Same completion
  /// contract as [clear]: a caller awaiting a frame on a stream that goes away
  /// while the gate is shut must not wait forever.
  List<QueuedAppFrame> dropStream(String streamId) {
    final dropped = <QueuedAppFrame>[];
    for (final ch in _channels) {
      final q = _queues[ch]!;
      var freed = 0;
      q.removeWhere((f) {
        if (f.streamId != streamId) return false;
        dropped.add(f);
        freed += f.plaintextBytes;
        return true;
      });
      _queuedBytes[ch] = _queuedBytes[ch]! - freed;
      if (q.isEmpty) blockedSince.remove(ch);
    }
    _complete(dropped);
    return dropped;
  }

  /// Resume the drain loop. Safe to call at any time: there is only ever one
  /// loop, and a call while it is already running is a no-op.
  void kick() {
    unawaited(_run());
  }

  void _complete(List<QueuedAppFrame> frames) {
    for (final f in frames) {
      if (!f.done.isCompleted) f.done.complete();
    }
  }

  bool _fits(QueuedAppFrame f) {
    final need = f.plaintextBytes + kSealOverheadBytes;
    final w = window;
    final cap = socketCap;
    final chUnacked = unacked(f.channel);
    final total = totalUnacked();
    // The "nothing outstanding" arms are the deadlock guards: a frame larger
    // than a limit still goes when no bytes are in flight, or it could never go
    // at all.
    final chOk = w == null || chUnacked == 0 || chUnacked + need <= w;
    final sockOk = cap == null || total == 0 || total + need <= cap;
    return chOk && sockOk;
  }

  /// The next frame to write: control before preview, always, so a preview
  /// backlog never delays a control frame — and a blocked control head never
  /// delays preview.
  QueuedAppFrame? _pick() {
    for (final ch in _channels) {
      final q = _queues[ch]!;
      if (q.isEmpty) {
        blockedSince.remove(ch);
        _stallWarned.remove(ch);
        continue;
      }
      if (_fits(q.first)) {
        blockedSince.remove(ch);
        _stallWarned.remove(ch);
        return q.first;
      }
      _noteBlocked(ch);
    }
    return null;
  }

  void _noteBlocked(String ch) {
    final since = blockedSince[ch] ??= DateTime.now();
    if (_stallWarned.contains(ch)) return;
    final stalledMs = DateTime.now().difference(since).inMilliseconds;
    if (stalledMs < kWindowStallWarnMs) return;
    _stallWarned.add(ch);
    _log?.call(
      'send gate stalled on $ch for ${stalledMs ~/ 1000}s: '
      'unacked=${unacked(ch)} totalUnacked=${totalUnacked()} '
      'queued=${_queues[ch]!.length} frame(s)/${_queuedBytes[ch]} bytes',
    );
  }

  Future<void> _run() async {
    if (_draining) return;
    _draining = true;
    try {
      while (true) {
        if (hold) return;
        final f = _pick();
        // Idle, or every head is gate-blocked; credit() and uncharge() kick
        // the loop again once the window reopens.
        if (f == null) return;
        _queues[f.channel]!.removeAt(0);
        _queuedBytes[f.channel] = _queuedBytes[f.channel]! - f.plaintextBytes;
        int? n;
        try {
          // The await is also the yield that lets a session frame written
          // meanwhile land between two app frames rather than behind all of
          // them.
          n = await _sink(f);
        } catch (_) {
          // A sink that throws must not strand the frames behind it, nor leave
          // this frame's `done` pending forever.
          n = null;
        }
        if (n != null) _sent[f.channel] = _sent[f.channel]! + n;
        if (!f.done.isCompleted) f.done.complete();
      }
    } finally {
      _draining = false;
    }
  }
}
